File: credit.c
#include<stdio.h>
#include<cs50.h>
int main(void)
{
    long n = get_long("Number:");
    long sum = 0;
    long place = 1;
    long a;
    int digit1;
        int digit2 = 0;

    while(true){
        a = n%10;
        if (place%2 == 0){
            int b = 2*a;
            if (b>9){
                sum += b%10 + b/10 ;
            }
            else{
                sum+= 2*a;
            }
        }
        else if (place%2 == 1)
        {
            sum = sum+a;
        }
        if (n/100 ==0 &&  n/10 != 0)
        {
            digit2 = a;
        }
        n = n/10;
        place +=1;
        if (n==0)
        {
            digit1 = a;
            break;
        }
        }
        if (sum%10 == 0){
            place -=1;
        if (digit1==4 && (place==13 || place==16))
        {
                printf("VISA\n");
        }

        else if (place==16)
        {
            if (digit1==5 && (digit2>=1 && digit2<=5))
            {
                printf("MASTERCARD\n");
            }
            else {
            printf("INVALID\n");
            }
        }

        else if (place==15)
        {
            if (digit1==3 && (digit2==4 || digit2==7))
            {
                printf("AMEX\n");
            }
            else {
            printf("INVALID\n");
            }
        }
        else {
            printf("INVALID\n");
            };
        }
        else{
            printf("INVALID\n");
        }
}    

File: pset6/credit/credit.py
from cs50 import get_string
import re
def main():
    cardno = get_string("Number: ")
    if luhns_algo(cardno) != True:
        print("INVALID\n")
        return
    else:
        if len(cardno)== 15:
            if re.search('^3[47].',cardno):
                print("AMEX\n")
        elif len(cardno) == 16:
            if re.search('^5[1-5].',cardno):
                print("MASTERCARD\n")
            elif re.search('^4.', cardno):
                print("VISA\n")
        elif len(cardno) == 13:
            if re.search('^4.', cardno):
                print("VISA\n")
        else:
            print("INVALID\n")
        

    
    
def luhns_algo(a):
    summ = []
    dumsum = 0
    for i in range(len(a)-1, -1 ,-1):
        if i%2 == 0:
            temp = int(a[i])
            temp *=2
            summ += str(temp)
        else:
            dumsum+= int(a[i])
    for i in summ:
        for j in i:
            dumsum+= int(j)

        return True
    else:
        return False
if __name__ == "__main__":
    main()